neasqc_wp61/models/quantum/beta/beta_2.py

```diff
@@ -0,0 +1,150 @@
+        x2 : np.array
+            Second of the vectors between which to compute the distance.
+        
+        Returns
+        -------
+        float
+            Quantum distance between the two vectors.
+        """
+        x1 = normalise_vector(x1)
+        x1 = pad_vector_with_zeros(x1)
+        x2 = normalise_vector(x2)
+        x2 = pad_vector_with_zeros(x2)
+        return qd(x1, x2).compute_quantum_distance()
+    
+    def initialise_cluster_centers(
+        self, random_initialisation = True,
+        seed = 30031935
+    )-> None:
+        """
+        Initialise the cluster centers for the algorithm.
+
+        Parameters
+        ----------
+        random_center_initializer : bool
+            If True randomly initialises the cluster centers.
+            If False, uses the K-Means++ algorirthm to initialise
+            cluster center. More info can be found in pycluster docs
+            https://pyclustering.github.io/docs/0.8.2/html/index.html.
+        seed : int
+            Seed to use when random_center_intializer = True.
+        """
+        if random_initialisation == True:
+            self.initial_centers = random_center_initializer(
+                self.x_train, self.k, random_state = seed
+            ).initialize()
+        else:
+            self.initial_centers = kmeans_plusplus_initializer(
+                self.x_train, self.k).initialize()
+
+    def train_k_means_algorithm(self)-> None:
+        """
+        Instantiate and train k-means algorithm. Assign the clusters, 
+        centers and n_iterations as attributes of the class.
+        """
+        self.k_means_instance = kmeans(
+            self.x_train, self.initial_centers,
+            tolerance = self.tolerance, observer = self.observer,
+            metric = self.metric, itermax = self.itermax
+        )
+        self.k_means_instance.process()
+        self.clusters = self.observer._kmeans_observer__evolution_clusters
+        self.centers = self.observer._kmeans_observer__evolution_centers
+        self.n_its = len(self.clusters)
+    
+    def compute_wce(self)-> None:
+        """
+        For each iteration of the algorithm training process, compute the within cluster
+        sum of errors:  
+        $[error=\sum_{i=0}^{N}quantum_distance(x_{i}-center(x_{i}))$].
+        """
+        self.total_wce = []
+        for i in range(self.n_its):
+            wce = 0
+            for j in range(self.k):
+                for item in self.clusters[i][j]:
+                    wce += self.wrapper_quantum_distance(
+                        self.x_train[item], self.centers[i][j]
+                    )
+            self.total_wce.append(wce)
+
+    def compute_predictions(self)-> None:
+        """
+        For each iteration of the algorithm training process, compute 
+        the predictions of the trainining instances, i.e.,
+        what the index of their closest cluster is. 
+        """
+        self.predictions = [
```

Comment:
What are these lines trying to accomplish and why?

Reply:
The library that we are using only computes the prediction for the last iteration of the algorithm. This function computes the predictions for all the iterations during the training process

Reply:
You're creating a list of lists of None.
I don't understand why you'd want to do that.
Could you explain why you need a list of lists of None?

I you really want to create a list of lists of None, there are more optimised ways to do so (you can lookup things related to repeat).

Otherwise you can just create it in list comprehension by changing a few of the things you have below, no need to create a big None object before populating it, you can create it directly!

Reply:
The reason why I need a list of None is because I need to I will add elements to it in a "disordered" way. In the loop in lines 149-152, I will add elements so that for example in the first iteration of the loop, I add what would be the fourth element of the list. After looking up here : 

https://stackoverflow.com/questions/50835478/a-way-to-insert-a-value-in-an-empty-list-in-a-specific-index

it seems that the only way of doing it is with a list of None or with a dictionary. I decided to use a list for the sake of consistency with the rest of the models. 

I changed the way I create it, so that it is slightly more optimised. I looked up the repeat function of itertools library (I am not sure if u were referring to it in your suggestion) but the code looks very similar to the solution I came up with, so I decided not to use itertools in order to avoid importing the library.